### PR TITLE
feat: generate solar profiles during HIFLD grid-building

### LINKED ISSUE #248: Add a module to create profiles for solar thermal generators
# :rocket:

- [ ] Is your feature request essential for your project?

### Describe the workflow you want to enable
Currently, for the HIFLD grid, we're excluding solar thermal generators, since we only have a profile generation process that applies to solar photovoltaic. However, if we can come up with reasonable profiles for for solar thermal, we can include them. This represents approximately 5% of the total solar capacity (1.8 GW of 37.8 GW total from EIA Form 860 for 2019).

### Describe your proposed implementation
Looking at the documentation for NREL's SAM tool (https://nrel-pysam.readthedocs.io/en/master/Models.html), it seems they have several solar thermal models, of which `TcsgenericSolar` seems like the most generic. We should be able to use this module via PySAM in a similar way as we currently use their PV model (`PVWattsv7`).

### LINKED ISSUE #254: Update solar profile generation to reflect that nrel-pysam can now handle leap years
# :rocket:

- [ ] Is your feature request essential for your project?

### Describe the workflow you want to enable
I wish that `prereise.gather.solardata.nsrdb.sam` were simplified, to reflect changes in NREL-PySAM v2.2.2: "Leap year 2/29 date is now acceptable".

### Describe your proposed implementation
Refactor the `retrieve_data_blended` and `retrieve_data_individual` functions to avoid using two distinct datetime arrays (the real datetimes and the everything-but-leap-day datetimes), and remove the `generate_timestamps_without_leap_day` helper function. Inputs/outputs should remain the same, with the exception that the leap day power profiles will change to reflect actual February 29th weather, not just a copy of February 28th.

### PR DESCRIPTION
**Pull request by danielolsen**
### Purpose
- Generate solar profiles at the same time as we build the HIFLD grid, so that we can leverage the plant-specific attributes of each solar farm to generate better profiles (partial fulfillment of #227).
- Add a module to generate profiles for concentrating solar power (CSP) plants, which represent about 5% of total solar capacity (closes #248).
- ~Simplify solar profile generation process, now that SAM can handle leap years.~ EDIT: this no longer appears possible, see https://github.com/Breakthrough-Energy/PreREISE/issues/254#issuecomment-1007800590
- Update the version of NREL's PySAM used, for compatibility with Python 3.10, and test using Python 3.10.

### What the code is doing
- Within `Pipfile`, `Pipfile.lock`, and `requirements.txt`: updating the version of PySAM. Then, we add Python 3.10 to the test matrix in `.github/workflows/test.yml`.
- Within `const.py`, we add references to the tables of 'extra' EIA data in the blob storage.
- Within `generators.py`, we make sure that the plant index gets named (unrelated to main objectives).
- We add a new module `prereise.gather.griddata.hifld.data_process.profiles`, which takes the 'extra' data table for solar, joins it with the solar plants from the plant dataframe, and passes this (with some pre-processing) to individualized profile generation function introduced in #247. Most of this new code is copy-and-pasted from the usage examples in that PR.
- Within `orchestration.py`, we call the new function from `prereise.gather.griddata.hifld.data_process.profiles`, and then save the resulting profiles to CSV at the same time as we save the rest of the Grid CSVs.
- Within `nrel_api.py`:
  - we add a method that converts the `Psm3Data` data structure back to the file format expected by SAM. For whatever reason, the way that weather data are attached to a `PySAM.PVWatts` object is different than how they're attached to a `PySAM.TcsgenericSolar` object: `TcsgenericSolar` needs a filename for a file formatted in a particular way.
  - we make the downloading more resilient to server-side errors.
- Within `sam.py`:
  - ~we remove the leap-year-specific logic, now that SAM handles lear years properly.~
  - we query the full year's weather data during leap years, since SAM can take it as input, but we still need to manually copy the Feb 28th solar data to leap days, since SAM doesn't produce power output for Feb 29th.
  - we rename the PV-profile-generating function from `calculate_power` to `calculate_power_pv`, and add a parallel `calculate_power_csp` function, which saves weather data into a temporary location and then tells the CSP model to load it from this location (it's ugly, but I don't know if there's a better way).
  - within `retrieve_data_individual`, we look at each plant's prime mover code to determine whether to calculate profiles using the PV model or the CSP model.
  - miscellaneous docstring fixes.

### Testing
CSP profile generation code has been manually tested. ~Manual testing of profiles for the full solar fleet is in progress, should be done by this afternoon (data are being re-downloaded from NREL with leap days).~ EDIT: tested manually, the solar profiles are generated and saved and match the number of solar plants in the plant CSV table.

### Usage Example/Visuals
```python
from prereise.gather.griddata.hifld import create_csvs
create_csvs(
    DESIRED_OUTPUT_CSV_DIRECTORY,
    YOUR_NREL_EMAIL,
    YOUR_NREL_API_KEY,
    {"year": 2020, "cache_dir": YOUR_CACHE_DIR},
)
```

### Time estimate
1 hour. None of the individual pieces are too complex, but there are a lot of moving parts here. Assuming that y'all approve of the overall logic flow, I'll submit a follow-up PR to add wind profile generation to the orchestration script as well (or I can add it to this PR).

### CODE REVIEW

File: prereise/gather/griddata/hifld/orchestration.py [hunk 1]
```diff
@@ -6,13 +6,18 @@
 from prereise.gather.griddata.hifld.const import powersimdata_column_defaults
 from prereise.gather.griddata.hifld.data_process.demand import assign_demand_to_buses
 from prereise.gather.griddata.hifld.data_process.generators import build_plant
+from prereise.gather.griddata.hifld.data_process.profiles import build_solar
 from prereise.gather.griddata.hifld.data_process.transmission import build_transmission
 
 
-def create_csvs(output_folder):
+def create_csvs(output_folder, nrel_email, nrel_api_key, solar_kwargs={}):
```
Comment (danielolsen) [inline prereise/gather/griddata/hifld/orchestration.py:13]: From the user perspective, do we want to enable the user to generate the grid without generating the profiles? If so, we could add a flag input, and add default `None` values for the profile-specific input parameters.

Comment (danielolsen) [reply prereise/gather/griddata/hifld/orchestration.py:13]: Alternatively, we may want the user to be able to create the full Grid CSVs, and then use these for multiple years. We could break `create_csvs` into a grid-step and a profiles step, or allow the user to pass a list of years to have everything handled automatically.

**Comment by danielolsen** [comment]
There's another branch `daniel/hifld_wind_profiles` which adds the wind profile creation step as well, see https://github.com/Breakthrough-Energy/PreREISE/compare/daniel/hifld_solar_profiles...daniel/hifld_wind_profiles.

**Comment by danielolsen** [comment]
> There's another branch `daniel/hifld_wind_profiles` which adds the wind profile creation step as well, see [daniel/hifld_solar_profiles...daniel/hifld_wind_profiles](https://github.com/Breakthrough-Energy/PreREISE/compare/daniel/hifld_solar_profiles...daniel/hifld_wind_profiles).

Similarly, there's a branch to add hydro profiles, see https://github.com/Breakthrough-Energy/PreREISE/compare/daniel/hifld_solar_profiles...daniel/hifld_hydro_profiles.

File: prereise/gather/solardata/nsrdb/nrel_api.py [hunk 2]
```diff
@@ -71,6 +71,19 @@ def to_dict(self):
         )
         return result
 
+    def to_sam_weather_file_format(self):
+        """Convert the data to the format expected by nrel-pysam for local files. See
+        https://developer.nrel.gov/docs/solar/nsrdb/psm3-download/.
+
+        :return: (*list*) -- a list of lists which can be passed to
```
Comment (jenhagg) [inline prereise/gather/solardata/nsrdb/nrel_api.py:78]: Technically I think `metadata_values` is a tuple, but would expect that's fine.

File: prereise/gather/solardata/nsrdb/sam.py [hunk 3]
```diff
@@ -1,7 +1,12 @@
+import csv
+import os
+import tempfile
+
 import numpy as np
 import pandas as pd
-import PySAM.Pvwattsv7 as PVWatts
+import PySAM.Pvwattsv8 as PVWatts
```
Comment (jenhagg) [inline prereise/gather/solardata/nsrdb/sam.py:7]: Should L#38 (in `calculate_power_pv`) still be using `"pvwattsv7"`?

File: prereise/gather/griddata/hifld/orchestration.py (hunk 1 above)
Comment (jenhagg) [reply prereise/gather/griddata/hifld/orchestration.py:13]: Agree it makes sense to be able to create the grid and profiles separately. Assuming this isn't something that would be used often, I might leave it at something like `create_grid` and `create_profiles`, or make `create_csvs` a wrapper around those if it's convenient.

File: prereise/gather/solardata/nsrdb/sam.py [hunk 4]
```diff
@@ -225,26 +234,36 @@ def retrieve_data_individual(
...
-                "array_type": plant_array_types.loc[plant_id],
-            }
-            if plant_pv_dict["array_type"] == 0:
-                plant_pv_dict["tilt"] = series["Tilt Angle"]
-            pv_dict = {**default_pv_parameters, **plant_pv_dict}
-            power = calculate_power(solar_data, pv_dict)
-            if leap_day is not None:
-                power = np.insert(power, leap_day, power[leap_day - 24 : leap_day])
-
+            if series["Prime Mover"] == "PV":
```
Comment (jenhagg) [inline prereise/gather/solardata/nsrdb/sam.py:245]: Could we add a comment to define the various prime mover codes?

Comment (danielolsen) [reply prereise/gather/solardata/nsrdb/sam.py:245]: Done.

File: prereise/gather/griddata/hifld/orchestration.py (hunk 1 above)
Comment (danielolsen) [reply prereise/gather/griddata/hifld/orchestration.py:13]: Done. **EDIT**: The refactor has also been tested and behaves as expected when calling the `create_csvs` wrapper function.

File: prereise/gather/solardata/nsrdb/sam.py (hunk 3 above)
Comment (danielolsen) [reply prereise/gather/solardata/nsrdb/sam.py:7]: Updated.

**Review by kasparm — APPROVED** [review]
Logic looks good.
Solar profile and grid-building works.
Thank you!

**Pull request merged**